### ISSUE TITLE
Allow the container to run as non-root, with a readonlyrootFS

### ISSUE THREAD

**Issue by davralin** [librespeed/speedtest#682]
## Description
As mentioned from my comment in #663 , the container runs with very high privileges, which stops it from running in environments where higher levels of security is needed/wanted.

Most of the things can be fixed, with some refactoring of both dockerfiles and the entrypoint, however:
- It will be a breaking change, as it would have to change the listening port in apache to be something above 1024.
- That will also break the intended (?) purpose of the container being available on default tcp/80 with `network_mode: host` (network_mode: host will still work, but the container will listen to (for instance) tcp/8080 instead)
- Something needs to chown the datadir with the sqlite-database, and it can't be the container, because that operation will require root-privileges

You mention `Given this configuration, no further adjustments are needed to meet our security goals.` - does this mean you don't want to reconfigure the containers so that they run as non-root, with a readonlyrootFS?
I'd hate to work on a PR, just for it to be declined.

## Why it should be implemented
People who don't want root-containers, with full read/write-access to it's own container's file system.

**Comment by sstidl** [comment]
You can go ahead. I understand your intentions better now. As far as I see it there's no need to run in host mode but maybe that's before my time here in the project.
Please try to be clear what your changes are when coding them. 
I will review and comment if something is unclear.

Thank you very much for the effort and input and time!

**Comment by scyto** [comment]
It is a common misconception that running root in the container is the same as running root on the host *it isn't* whether you run as root in the container or some other GID/UID you have the same host capabilities

the port issue can be fixed with the existing envar IMO

**Comment by davralin** [comment]
> Please try to be clear what your changes are when coding them. I will review and comment if something is unclear.

My contributions will most likely only be around the Dockerfile's and the entrypoint, and related documentation for that.
I'm no coder :-)

At a high-level, my plan is to remove most of the things from entrypoint, and move it to a multi-part Dockerbuild.
I think I'm going to deprecate the webport entirely, port-mappings should be sufficient, so it makes no sense to have to configure something twice, in my opinion.
I'm not sure yet, how I can fix all the custom copying in entrypoint.sh in a readonlyrootFS (because that's not possible, after all :-) ) - so that might be something that someone else needs to fix.

**Comment by scyto** [comment]
we need the ability to change the port by envar, some of us want to run these on macvlans, or host where port publishing is not required

Wehn you say multipart docker build i want to make sure you are not talking about building the image at run time - just it is multipart for the dcokerfile to produce a static image for the public registry?

**Comment by NovaFox161** [comment]
I don't think this should be closed, adding support to run as a rootless container is a major security improvement.

**Comment by scyto** [comment]
> > I don't think this should be closed, adding support to run as a rootless container is a major security improvement.


you are totaly confused  (as is much of the internet) what running as root means, when a container runs with a UID of 0 this means thats its UID inside the container, it DOES NOT HAVE ROOT ACCESS to your docker host, infact if you run a container with a different UID/GID the two containers have EXACTLY the same CAPs as mediated by the docker daemon

the only plance the UID/GID is used is for accessing bind mount files, and what you see there is not the container being root but the inherent insecurity of linux file permissions where any process can claim what bit mask it needs, and if you have given it access to a bind mount you already decided you want to have RW or RO permission

tl;dr your request is uneccesary and pointless and doesn't improve security, running in rootless is something entirely different - that means the daemon doesn't run as root (thats great, but major functionality is lost when doing that - esp host networking access) - you could try running in podman if you prefer

**Comment by davralin** [comment]
> you are totaly confused (as is much of the internet)

lol.

**Comment by thecodygriffin** [comment]
While acknowleding this is an open source project and anyone is free to use or not use it, this stance is how you lose potential users and contributors.

As self-hosting becomes more popular, more people will become aware of how running with root privileges is not the greatest idea. I refuse to run docker services as root and immediately dismiss projects that do not have a rootless, arbitrary user option as not yet ready for widespread usage.

Good luck with this project, but I'd recommend that others stay far away as long as this is the accepted stance of the project.

**Comment by sstidl** [comment]
> While acknowleding this is an open source project and anyone is free to use or not use it, this stance is how you lose potential users and contributors.
> 
> As self-hosting becomes more popular, more people will become aware of how running with root privileges is not the greatest idea. I refuse to run docker services as root and immediately dismiss projects that do not have a rootless, arbitrary user option as not yet ready for widespread usage.
> 
> Good luck with this project, but I'd recommend that others stay far away as long as this is the accepted stance of the project.

While it's a good practice to run services not as root in a container, it's not as bad as you described it above. As this project has not many maintainers, I invite anyone who has the time to change the need of root in the container and submit a PR.

**Comment by davralin** [comment]
> Good luck with this project, but I'd recommend that others stay far away as long as this is the accepted stance of the project.

https://github.com/librespeed/speedtest-go

Looks like this is a better fit.
